app/src/main/java/com/artemzin/qualitymatters/network/HostSelectionInterceptor.java

```diff
@@ -0,0 +1,54 @@
+package com.artemzin.qualitymatters.network;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import okhttp3.HttpUrl;
+import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
+
+/**
+ * Such implementation allows us easily change base url in the integration and functional tests!
+ */
+public class HostSelectionInterceptor implements Interceptor {
+
+    @Nullable
+    private volatile String baseUrl;
+
+    @Override @NonNull
+    public Response intercept(@NonNull Chain chain) throws IOException {
+        Request request = chain.request();
+        final String baseUrl = this.baseUrl;
+
+        if (baseUrl != null) {
+            request = modifyRequest(request, baseUrl);
```

Comment:
Wait wait wait then :)

Please check https://github.com/square/retrofit/pull/1652 and https://github.com/artem-zinnatullin/qualitymatters/issues/111

The idea is that `Retrofit` has immutable `baseUrl` so we can recreate `Retrofit` and switch it to new url.

Recreating `OkHttpClient` is not that great, we use it for many things and using one instance for everything is great because it reuses same connection pool and we don't allocate memory and resources for using multiple clients